### PR TITLE
Add pipeline which automates testing and releasing the gem

### PR DESCRIPTION
**Pull request by raphael-m**
We want to automatically test, build and push the gem to RubyGems when merging changes to the main branch. This PR prepares the Github Actions workflow to accomplish this.

The pipeline currently has two workflow files, the `tests.yml` which
* runs on pull requests to the main branch and when pushing / merging to the main branch itself
* has a matrix test strategy which tests on Ruby 2.5, 2.6 and 2.7
* runs `rake spec` to run the gem tests

and the `release.yml` which
* runs upon being manually triggered
* adds a tag to Github with the current gem version (in `lib/keycloak_oauth/version.rb`) and
* publishes the gem to RubyGems

Additional notes:
* Unfortunately it's currently possible to run the release workflow without the tests having passed. I did not find a way to require this (because the release workflow must be defined in a separate file).
* The manual action trigger does appear in Github Actions once it's merged to the main branch.
* We can't use `rake release` because it requires some Github configuration like user name, email etc. to be set up. This is why we separately tag to Github and then use `gem build` and `gem push` as guided in the Github Actions starter template (https://github.com/actions/starter-workflows/blob/d7ac62140faf23b67c29e892d4ce68342eb09609/ci/gem-push.yml#L33)

### CODE REVIEW

File: .github/workflows/release.yml [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+    # needs: test
+    if: github.ref == 'refs/heads/main'
+    steps:
+      - uses: actions/checkout@v2
+      - uses: ruby/setup-ruby@v1
+        with:
+          ruby-version: 2.7
+      - name: Tag on Github
+        run: |
+          # Todo: maybe there's a better way to get the version
```
Comment (raphael-m) [inline .github/workflows/release.yml:21]: If you have some advice here, let me know

File: .github/workflows/tests.yml [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+    name: Test on Ruby ${{ matrix.ruby }}
+    steps:
+      - uses: actions/checkout@v2
+      - uses: ruby/setup-ruby@v1
+        with:
+          ruby-version: ${{ matrix.ruby }}
+      - uses: actions/cache@v1
+        with:
+          path: vendor/bundle
+          key: bundle-${{ matrix.ruby }}-${{ hashFiles('**/*.gemspec') }}
```
Comment (crackofdusk) [inline .github/workflows/tests.yml:24]: I think you would also want to invalidate the cache on `Gemfile.lock` changes.

File: .github/workflows/tests.yml [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+          restore-keys: bundle-${{ matrix.ruby }}
+      - run: bundle install --jobs 4 --retry 3 --path vendor/bundle
```
Comment (crackofdusk) [inline .github/workflows/tests.yml:26]: Does `nproc` work on Github? I find some allusions to it timing out. There's a workaround as well: https://github.community/t/number-of-cores-as-env-context-variable/17396

Then again, gems are cached, so it's not big deal even if 4 is a suboptimal number of jobs.

File: .github/workflows/release.yml (hunk 1 above)
Comment (crackofdusk) [reply .github/workflows/release.yml:21]: I can only think of an indirect way that's higher level:

```ruby
puts Gem::Specification::load("keycloak_oauth.gemspec").version
```

File: .github/workflows/tests.yml (hunk 3 above)
Comment (crackofdusk) [reply .github/workflows/tests.yml:26]: Also check the deprecation warning:

> [DEPRECATED] The `--path` flag is deprecated because it relies on being remembered across bundler invocations, which bundler will no longer do in future versions. Instead please use `bundle config set path 'vendor/bundle'`, and stop using this flag

**Review by crackofdusk — COMMENTED** [review]
> Unfortunately it's currently possible to run the release workflow without the tests having passed. I did not find a way to require this (because the release workflow must be defined in a separate file).

Ideas:

- duplicate the test job?
- Have a single workflow for testing and releasing. Tagging is in a separate workflow. Conditionally release on the tag event. 


(Inspired by https://github.community/t/sharing-steps-between-jobs-and-jobs-between-workflows/17751/3,  https://stackoverflow.com/a/58459586. There's also this, I don't know if it helps: https://github.community/t/reusing-sharing-inheriting-steps-between-jobs-declarations/16851/17)

File: .github/workflows/release.yml [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+
+jobs:
+  release:
+    name: Release to RubyGems
+    runs-on: ubuntu-latest
+    # Ideally, we would require successful tests before
+    # allowing to publish. Because the tests need to be
+    # in a separate workflow file, we can't reference it
+    # here.
+    # needs: test
```
Comment (ahorincar) [inline .github/workflows/release.yml:12]: But we could add a `run` block here to run the tests, no? Or am I misunderstanding something?

Comment (andyundso) [reply .github/workflows/release.yml:12]: I think this thread tackles the issue fairly well: https://github.community/t/depend-on-another-workflow/16311/3. The key point: Because each PR needs to have passing tests, you know that master will implicitly pass as well.

Comment (raphael-m) [reply .github/workflows/release.yml:12]: Well the PR might have passing tests, but if merging occurs (in case a branch is not fully rebased to master), this does not imply that the tests also pass for master. Do you think we can safely ignore that @andyundso ?

Technically, we could add another `run` block to run the tests, but then we would be forced to basically duplicate the test workflow, and that hindered me from doing so.

File: .github/workflows/tests.yml (hunk 3 above)
Comment (raphael-m) [reply .github/workflows/tests.yml:26]: Thanks, I'll replace the `--path` argument.

I'll also reduce the number of jobs to `2` because apparently the Github-hosted virtual machines all have two cores (https://docs.github.com/en/free-pro-team@latest/actions/reference/specifications-for-github-hosted-runners).

File: .github/workflows/release.yml [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+    if: github.ref == 'refs/heads/main'
+    steps:
+      - uses: actions/checkout@v2
+      - uses: ruby/setup-ruby@v1
+        with:
+          ruby-version: 2.7
+      - name: Tag on Github
+        run: |
+          VERSION=$(echo "puts Gem::Specification::load('keycloak_oauth.gemspec').version" | ruby)
```
Comment (crackofdusk) [inline .github/workflows/release.yml:21]: You can also do this:

```suggestion
          VERSION=$(ruby -e "puts Gem::Specification::load('keycloak_oauth.gemspec').version")
```

It's just for your information. Both versions are fine.

File: .github/workflows/release.yml (hunk 1 above)
Comment (raphael-m) [reply .github/workflows/release.yml]: I'm now using this, thanks

File: .github/workflows/release.yml (hunk 5 above)
Comment (raphael-m) [reply .github/workflows/release.yml]: I like `ruby -e` so I committed your suggestion

**Review by crackofdusk — APPROVED** [review]

File: .github/workflows/release.yml (hunk 4 above)
Comment (andyundso) [reply .github/workflows/release.yml:12]: You can force this within the Github settings that every PRs branch needs to be up-to-date with master before it can be merged.

Otherwise, you can still port the pipeline to Semaphore, if deduplication is a concern.

Comment (raphael-m) [reply .github/workflows/release.yml:12]: I added a branch protection rule to enforce up-to-date PRs before merging.

**Review by ahorincar — APPROVED** [review]
